### PR TITLE
Third-party exporter loading framework

### PR DESCRIPTION
**Pull request by prydin**
Exporters are packaged in separate jar-files and loaded through the ```ServiceLoader``` mechanism. Exporters need to specify the implementation under META-INF according to general SPI guidelines.

Shading of any references to the OpenTelemetry API are handled at load time, keeping the build scripts of the exporters clean.

Currently, there's a ```exportersupport``` module for sharing factory and config interfaces between the agent and exporters. This should be moved to opentelemetry-java at some point.

### CODE REVIEW

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  @Override
+  public int getInt(final String key, final int defaultValue) {
+    final String s = Config.getSettingFromEnvironment(prefix + "." + key, null);
+    if (s == null) {
+      return defaultValue;
+    }
+    return Integer.parseInt(s); // TODO: Handle format errors gracefully?
+  }
+
+  // @Override
```
Comment (trask) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java]:
```suggestion
```

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  @Override
+  public long getLong(final String key, final long defaultValue) {
+    final String s = Config.getSettingFromEnvironment(prefix + "." + key, null);
+    if (s == null) {
+      return defaultValue;
+    }
+    return Long.parseLong(s); // TODO: Handle format errors gracefully?
+  }
+
+  // @Override
```
Comment (trask) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java]:
```suggestion
```

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  @Override
+  public boolean getBoolean(final String key, final boolean defaultValue) {
+    final String s = Config.getSettingFromEnvironment(prefix + "." + key, null);
+    if (s == null) {
+      return defaultValue;
+    }
+    return Boolean.parseBoolean(s); // TODO: Handle format errors gracefully?
+  }
+
+  // @Override
```
Comment (trask) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java]:
```suggestion
```

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    this.prefix = prefix;
+  }
+
+  // @Override
+  @Override
+  public String getString(final String key, final String defaultValue) {
+    return Config.getSettingFromEnvironment(prefix + "." + key, defaultValue);
+  }
+
+  // @Override
```
Comment (trask) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java]:
```suggestion
```

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import io.opentelemetry.auto.exportersupport.ConfigProvider;
+
+public class DefaultConfigProvider implements ConfigProvider {
+  private final String prefix;
+
+  public DefaultConfigProvider(final String prefix) {
+    this.prefix = prefix;
+  }
+
+  // @Override
```
Comment (trask) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/DefaultConfigProvider.java]:
```suggestion
```

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/ShadingRemapper.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+  public ShadingRemapper(final Rule... rules) {
+    for (final Rule rule : rules) {
+      map.put(rule.from, rule.to);
+    }
+  }
+
+  @Override
+  public String map(final String internalName) {
+    final Map.Entry<String, String> e = map.floorEntry(internalName);
```
Comment (trask) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/ShadingRemapper.java]: 👍 I've never seen `floorEntry()` before, was wondering why `TreeMap` above

File: agent-tooling/src/test/groovy/io/opentelemetry/auto/tooling/ExporterLoaderTest.groovy [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package io.opentelemetry.auto.tooling
+
+
+import io.opentelemetry.auto.util.test.AgentSpecification
+import io.opentelemetry.sdk.OpenTelemetrySdk
+
+class ExporterLoaderTest extends AgentSpecification {
```
Comment (trask) [inline agent-tooling/src/test/groovy/io/opentelemetry/auto/tooling/ExporterLoaderTest.groovy]: 👍

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/TracerInstaller.java [hunk 8]
```diff
@@ -1,37 +1,86 @@
...
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import java.util.jar.Manifest;
 import lombok.extern.slf4j.Slf4j;
 
 @Slf4j
 public class TracerInstaller {
+  private static ExporterClassLoader exporterLoader;
```
[the remark this answers is not in the document]
Comment (prydin) [reply agent-tooling/src/main/java/io/opentelemetry/auto/tooling/TracerInstaller.java]: Not it's not. Removed.

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/ShadingRemapper.java (hunk 6 above)
Comment (prydin) [reply agent-tooling/src/main/java/io/opentelemetry/auto/tooling/ShadingRemapper.java]: Yeah, that's a great trick for checking if a string with a common prefix exists in a list. Probably a bit overkill for the small list we have now, but you never know if it's going to grow...

**Pull request merged**